feat(rt): add ProcessCredentialsProvider

lauzadis opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
Adds a ProcessCredentialsProvider which will be used to support [process credentials](https://docs.aws.amazon.com/sdkref/latest/guide/feature-process-credentials.html) resolution

## Issue \#
N/A

## Description of changes
This change is required to achieve feature parity with existing SDKs. It allows more customization on where the SDK gets users' credentials from.

Process credentials may be returned without an expiration field. In this case, we are expected to treat the credentials as non-expiring, and no credential refreshes should be attempted. To support this behavior, the concept of `NonExpiringCredentials` was added to the JSON deserializer used for credentials. If we receive non-expiring credentials, we set the expiration to `Instant.MAX_VALUE` to ensure they "never" expire. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

aajtodd reviewed (changes requested):
Looks pretty good overall! Couple fixes and suggestions

lauzadis commented on aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/auth/credentials/ProcessCredentialsProvider.kt:
```diff
@@ -0,0 +1,70 @@
+                resp.accessKeyId,
+                resp.secretAccessKey,
+                resp.sessionToken,
+                Instant.MAX_VALUE,
```
Yeah, I was going to use this initially, but Java's `Instant` class does not like taking Long.MAX_VALUE as input.

```
java.time.DateTimeException: Instant exceeds minimum or maximum instant
```

Also the epoch-seconds-date of Long.MAX_VALUE is sometime in 2262, but Instant.MAX_VALUE is [defined to be `1000000000-01-01T00:00Z`](https://docs.oracle.com/en/java/javase/13/docs/api/java.base/java/time/Instant.html#MAX) which is much further in the future.

lauzadis commented on aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/auth/credentials/JsonCredentialsDeserializer.kt:
```diff
@@ -128,3 +136,58 @@ internal suspend fun deserializeJsonCredentials(deserializer: Deserializer): Jso
+ * The difference between this and [deserializeJsonCredentials] is that process credentials _must_ provide a version field,
+ * the session token field is called `SessionToken` instead of `Token`, and the expiration field is optional.
+ */
+internal fun deserializeJsonProcessCredentials(deserializer: Deserializer): JsonCredentialsResponse {
```
I agree that the duplication is not great. Initially I followed your suggestion to only validate `Version` as 1 if it's present, but the specification says that the `Version` field is **required** for process credentials, so I wanted to enforce that and didn't see a way to keep it backwards-compatible with the current credentials providers.

Another issue I ran into when re-using the existing `deserializeJsonCredentials` is parsing the session token. The current credentials providers parse it from the `Token` key, but credentials providers provide it using the `SessionToken` key.

So the `Version` and `SessionToken` vs. `Token` issues made me lean towards creating an entirely new method instead. I guess it's a question of special-casing in the current deserializer function or making an entirely new one for this process credentials format.

lauzadis replied:
Yeah I would like to get the rest of the team's opinion on this. The only one concern I have with your suggestion (and what I initially implemented) is that we will not be **requiring** the `Version` field to be set, which is required by the specification:

> The Version key must be set to 1 (as an integer type).

aajtodd replied:
> I'd much prefer we were able to codegen these structs and deserializers just like model shapes

That would pretty much guarantee that they couldn't be shared (it also wouldn't allow custom logic like validating `version` or dealing with different versions).

> we also do this in [S3ErrorMetadata](https://github.com/awslabs/aws-sdk-kotlin/blob/230751e55e50afe97c263bcf2698202b959a18b0/services/s3/common/src/aws/sdk/kotlin/services/s3/internal/S3ErrorMetadata.kt#L45-L78)

Maybe I'm missing something but I'm not sure what the issue is with hand written deserializers? Even libraries like `kotlinx.serialization`, `Jackson`, and others allow you to write serde code by hand when the default annotations/bindings don't yield what you want. This seems entirely within the expected use to me.

I can live with the duplication it just irks me because the whole idea of having a sealed type was to deal with differences in JSON responses. It would seem the expected class of JSON responses is more dissimilar than we originally expected :(

lauzadis commented:
Question: should I add `@InternalSdkAPI` decorators to the new functions such as `executeCommand`? It's not really intended for SDK users and could be subject to breaking changes in the future (i.e if we want to return stdout **and** stderr instead of just one of them, or generalize it even more with a larger refactor)

aajtodd commented:
> Question: should I add `@InternalSdkAPI` decorators to the new functions such as `executeCommand`? It's not really intended for SDK users and could be subject to breaking changes in the future (i.e if we want to return stdout **and** stderr instead of just one of them, or generalize it even more with a larger refactor)


It's marked `internal` already so it isn't consumable by anything other than from this project/compilation unit.

lauzadis commented on aws-runtime/aws-config/jvm/src/aws/sdk/kotlin/runtime/auth/credentials/executeCommand.kt:
```diff
@@ -0,0 +1,41 @@
+        cmd.add("cmd.exe")
+        cmd.add("/C")
+    } else {
+        cmd.add("sh")
```
Passing the command to a shell rather than starting it ourselves gives a little bit more configurability for users. 

For example, there are some things that a direct `.exec` call does not support, but sending command through the shell does support:
- variable substitution
- squiggle (~) / bang (!) / globbing (*) support 
- see more here: https://stackoverflow.com/a/68250664

The specification would be completely satisfied by executing the user-supplied command directly, but this just adds more configurability. This is also how the other SDKs do it, so I think we should support features like variable substitution for compatibility with scripts used today.

ianbotsf replied:
This also alleviates us from the responsibility of parsing a single string into the command-and-args-array expected by `ProcessBuilder`. Correctly handling quoted and nested arguments is tricky.

lauzadis commented on aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/auth/credentials/JsonCredentialsDeserializer.kt:
```diff
@@ -29,6 +30,13 @@ internal sealed class JsonCredentialsResponse {
     ) : JsonCredentialsResponse()
 
+    // Credentials that don't expire
+    data class NonExpiringCredentials(
```
Yeah, that would work. It would require some extra checks in the other providers (which today expect `expiration` to be non-null).

aajtodd replied:
They don't really expect it though, they just pass it through to `Credentials` constructor which takes `Instant?` don't they?

lauzadis replied:
ImdsCredentialsProvider expects it to be non-null [when calculating the nextRefresh time](https://github.com/awslabs/aws-sdk-kotlin/blob/main/aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/auth/credentials/ImdsCredentialsProvider.kt#L99). I can add a non-null check here and it will behave the same

aajtodd commented on aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/auth/credentials/ProcessCredentialsProvider.kt:
```diff
@@ -0,0 +1,62 @@
+                resp.expiration ?: Instant.MAX_VALUE,
+                PROVIDER_NAME,
+            )
+            else -> {
```
style: unnecessary block `{}`

aajtodd reviewed (approved):
Looks great. Nice work. Have you tested this out with real CLI process on different OS (e.g. linux + windows)?

lauzadis commented:
> Looks great. Nice work. Have you tested this out with real CLI process on different OS (e.g. linux + windows)?

Yeah, I just finished manual testing on all platforms (mac, linux, windows) and it works everywhere. I tested for command successes and failures (where stderr message is expected) for both absolute command paths and resolving the command from the system's PATH.

lauzadis commented:
Changes since last review:
- Add a configurable max bytes read from stdout
- Add a configurable command timeout
- Handle spaces in command by wrapping the entire string in quotes
- Add security warning to the KDocs
- Prevent calling getCredentials() after close() in CachedCredentialsProvider

ianbotsf commented on aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/auth/credentials/CachedCredentialsProvider.kt:
```diff
@@ -53,8 +53,9 @@ public class CachedCredentialsProvider(
 
     private val cachedCredentials = CachedValue<Credentials>(null, bufferTime = refreshBufferWindow, clock)
+    private var isClosed: Boolean = false
 
-    override suspend fun getCredentials(): Credentials = cachedCredentials.getOrLoad {
+    override suspend fun getCredentials(): Credentials = if (!isClosed) cachedCredentials.getOrLoad {
```
**Correctness**: Access/mutation of `isClosed` should be thread-safe.

aajtodd replied:
See [HttpClientEngineBase](https://github.com/awslabs/smithy-kotlin/blob/main/runtime/protocol/http/common/src/aws/smithy/kotlin/runtime/http/engine/HttpClientEngine.kt#L45) for an example of using an atomic for this purpose.

**style**: This would also be clearer written as a normal function body at this point IMO:

```kotlin
suspend fun getCredentials(): Credentials {
    check(!isClosed) { ... }

    return cachedCredentials.getOrLoad { ... }
}
```

ianbotsf commented on aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/auth/credentials/CachedCredentialsProvider.kt:
```diff
@@ -65,9 +66,11 @@ public class CachedCredentialsProvider(
-    }
+    } else throw IllegalStateException("Credentials provider is closed")
 
     override fun close() {
+        isClosed = true
+        cachedCredentials.close()
         (source as? Closeable)?.close()
     }
```
**Suggestion**: Since we now track closed status, consider only invoking this logic if `isClosed == false`.

aajtodd replied:
**fix**: I'll take Ian's suggestion a bit further and say that we want to make close idempotent. If we are already closed just return.

ianbotsf commented on aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/auth/credentials/ProcessCredentialsProvider.kt:
```diff
@@ -4,37 +4,48 @@ import aws.smithy.kotlin.runtime.auth.awscredentials.Credentials
  * If the exit code is non-zero, log and throw an exception including the stderr from the command.
  *
+ * WARNING: This [CredentialsProvider] sources credentials from an external process.
+ * This can potentially be dangerous, so proceed with caution. Other credential providers should be preferred if at all possible.
```
**Nit**: "at all" is redundant.

ianbotsf commented on aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/auth/credentials/ProcessCredentialsProvider.kt:
```diff
@@ -4,37 +4,48 @@ import aws.smithy.kotlin.runtime.auth.awscredentials.Credentials
  *
+ * WARNING: This [CredentialsProvider] sources credentials from an external process.
+ * This can potentially be dangerous, so proceed with caution. Other credential providers should be preferred if at all possible.
+ * If using this option, you should make sure that the config file and any process or script files are as locked down as possible using
+ * security best practices for your operating system.
```
**Style**: This feels both verbose and accusatory (via the word "you"). Perhaps it would be clearer, friendlier, and still equally cautionary to say something like:

> If using this option, please secure the config file and process/script files using the best practices for your operating system.

lauzadis replied:
Sure, I can update the phrasing.

FWIW, this warning statement was taken directly from the [public documentation](https://docs.aws.amazon.com/sdkref/latest/guide/feature-process-credentials.html) and only slightly modified to reference process/script files in addition to the config file.

ianbotsf commented on aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/auth/credentials/ProcessCredentialsProvider.kt:
```diff
@@ -4,37 +4,48 @@ import aws.smithy.kotlin.runtime.auth.awscredentials.Credentials
         val (exitCode, output) = try {
-            executeCommand(credentialProcess, platformProvider.osInfo().family == OsFamily.Windows)
+            executeCommand("\"$credentialProcess\"", platformProvider, maxOutputLengthBytes, timeoutMillis)
         } catch (ex: Exception) {
```
**Question**: [Public documentation for this feature](https://docs.aws.amazon.com/sdkref/latest/guide/feature-process-credentials.html#feature-process-credentials-detail-path) already shows examples of putting the process name in quotes. Will this logic lead to two pairs of quotes around process names? Does that work correctly?

lauzadis replied:
It does not work properly, I will need to handle this in a better way

ianbotsf commented on aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/auth/credentials/ProcessCredentialsProvider.kt:
```diff
@@ -52,19 +63,10 @@ public class ProcessCredentialsProvider(
-            is JsonCredentialsResponse.Error -> {
-                throw CredentialsProviderException("Error parsing credentials from process: code=${resp.code}; ${resp.message}")
-            }
+            else -> throw CredentialsProviderException("Credentials response was not of expected format")
```
**Question**: The extensible credentials providers spec says we should make the error message available to the user. Why did we remove the error code and message from the thrown exception?

lauzadis replied:
This is a different error message that gets thrown during parsing, it was not modified since the last review.

The error message provided to the user is not removed, it's here: https://github.com/awslabs/aws-sdk-kotlin/pull/729/files#diff-3c4a3e2a9ebfd55820d9f8123b9ac08bebda66c723e49b7826302d73532f9047R54

ianbotsf commented on aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/config/CachedValue.kt:
```diff
@@ -59,4 +60,6 @@ internal class CachedValue<T> (
         return refreshed.value
     }
+
+    override fun close() { value = null }
```
**Correctness**: Wrap the `value` mutation in the mutex lock.

lauzadis replied:
Wrapping this mutation with the mutex makes the function suspending, which means it can't be called from a `Closeable` because the overridden `close()` function can't be suspending.

I do think a mutex is necessary here to ensure we are nullifying the value properly. So, should I remove the implementation of `Closeable` in `CachedCredentialsProvider` and make `close()` a suspending function?

lauzadis replied:
Actually that idea will not work, because [the credentials chain only closes `Closeable`s](https://github.com/awslabs/aws-sdk-kotlin/blob/main/aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/auth/credentials/CredentialsProviderChain.kt#L55)

ianbotsf replied:
I take it this concern is allayed if we go remove `Closeable` and go with `evict` instead?

lauzadis replied:
I removed `Closeable` and renamed the function, but the race condition still exists right?

Without a mutex, if one thread calls `evict()` while another tries to `get()` the value, it's not clear what will happen.
If I make `evict()` use a mutex, it cannot be called from `CachedCredentialsProvider`'s `close()`.

ianbotsf commented on aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/config/CachedValue.kt:
```diff
@@ -59,4 +60,6 @@ internal class CachedValue<T> (
         val refreshed = initializer().also { value = it }
         return refreshed.value
     }
+
+    override fun close() { value = null }
```
**Comment**: A bit outside the scope of this PR but the name of this class makes even less sense now that it's closeable. This class isn't really a "cached value"...it's a "caching value provider" or, more commonly, a "cache".

aajtodd replied:
**correctness**: I don't think this type should implement `Closeable`. It's not a resource and could theoretically be used again after "closing" it.  I'd probably just add a new function such as `evict()` that sets the expired value to `null`. Then we should just evict the cached value in the `close()` implementation of `CachedCredentialsProvider`

ianbotsf commented on aws-runtime/aws-config/jvm/src/aws/sdk/kotlin/runtime/auth/credentials/executeCommandJVM.kt:
```diff
@@ -0,0 +1,67 @@
+        val start = clock.now().epochMilliseconds
+
+        while (process.isAlive && clock.now().epochMilliseconds - start < timeoutMillis) {
+            reader.readLine()?.let { output.append(it) }
```
**Correctness**: This reads lines from the process output but writes inline to the `StringBuilder` thus munching all the newlines.

aajtodd replied:
Also how does this respect timeouts? At any rate it seems like it would just be easier to wrap the reader loop in a `withTimeout` block, e.g.:

```kotlin

val process = ...
val reader = ...

withTimeout(timeout) {

    while(process.isAlive) {
        reader.readLine()?.let { ... } // or whatever the appropriate read logic is

        // check output length, etc
    }
}

```

lauzadis replied:
It respects timeouts because `reader.readline()` does not block, it either returns a line of text or `null` if no text is available. So the timeout condition in the `while` loop is constantly being evaluated.

I think `withTimeout` is a clearer option though 👍 

I will also refactor to avoid consuming the newline characters

ianbotsf commented on aws-runtime/aws-config/jvm/src/aws/sdk/kotlin/runtime/auth/credentials/executeCommandJVM.kt:
```diff
@@ -0,0 +1,67 @@
+        val output = StringBuilder()
+        val start = clock.now().epochMilliseconds
+
+        while (process.isAlive && clock.now().epochMilliseconds - start < timeoutMillis) {
+            reader.readLine()?.let { output.append(it) }
+
+            if (output.length > maxOutputLengthBytes) {
+                throw RuntimeException("Process output exceeded limit of $maxOutputLengthBytes bytes")
+            }
+        }
```
**Question**: This loop seems like it could be fairly CPU-intensive. How does this perform on processes that take a long time to return (e.g., due to network calls)?

aajtodd commented on aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/config/CachedValue.kt:
```diff
@@ -29,34 +33,58 @@ internal class CachedValue<T> (
+    fun get(): T? {
+        check(!closed.value) { "value is closed" }
+
+        if (ref == null || isExpired(ref!!)) {
```
**correctness**: This is doing 3 atomic loads of a `ref`. Typically with atomics you'll want to load and store it into a local variable.

e.g.

```kotlin
check(!closed) { ... }
val curr = ref ?: return null
return isExpired(curr)
```

This ensures the value isn't updated from a different thread/coroutine in middle of a single function call.

aajtodd replied:
Moreover since this doesn't modify the value should this throw an exception or just return `null`? 🤔 

```kotlin
fun get(): T? {
    val curr = ref
    if (curr == null || !isClosed || isExpired(curr)) return null
    return curr.takeIf { !isClosed && it != null && !isExpired(it) }
}
```


Summon @ianbotsf.

ianbotsf replied:
I would argue that it calling `get` on a closed instance should throw an exception, not return `null`.

```kotlin
fun get(): T? {
    check(!closed.value) { "value is closed" }
    return ref?.takeIf { !isExpired(it) }
}
```

aajtodd commented on aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/config/CachedValue.kt:
```diff
@@ -29,34 +33,58 @@ internal class CachedValue<T> (
-    private fun isExpiredUnlocked(): Boolean {
-        val curr = value ?: return true
-        return clock.now() >= (curr.expiresAt - bufferTime)
+    fun isExpired(): Boolean {
```
**style**: this can be turned into a property now 


```kotlin
val isExpired: Boolean
    get() { ... }
```

ianbotsf reviewed (approved):
Looks good to me!

aajtodd commented on aws-runtime/aws-config/build.gradle.kts:
```diff
@@ -43,6 +43,9 @@ kotlin {
                 implementation(project(":aws-runtime:protocols:aws-json-protocols"))
+
+                // atomics
+                implementation("org.jetbrains.kotlinx:atomicfu:$atomicFuVersion")
```
fix:

This should apply the atomicfu gradle plugin as well (see [example](https://github.com/awslabs/smithy-kotlin/blob/main/runtime/io/build.gradle.kts#L6-L25) from smithy-kotlin)

```kotlin
// right after the copyright (buildscript blocks must come before anything else except imports)
buildscript {
    val atomicFuVersion: String by project
    repositories {
        mavenCentral()
    }

    dependencies {
        classpath("org.jetbrains.kotlinx:atomicfu-gradle-plugin:$atomicFuVersion")
    }
}

...
val coroutinesVersion: String by project
val atomicFuVersion: String by project

apply(plugin = "kotlinx-atomicfu")

...

```

aajtodd commented on aws-runtime/aws-config/jvm/src/aws/sdk/kotlin/runtime/auth/credentials/executeCommandJVM.kt:
```diff
@@ -0,0 +1,72 @@
+
+                output.append(buffer)
+                if (output.length > maxOutputLengthBytes) {
+                    throw RuntimeException("Process output exceeded limit of $maxOutputLengthBytes bytes")
```
**question**: Should this be `ClientException` or `CredentialsProviderException`?

lauzadis replied:
I changed it to `CredentialsProviderException` 👍

aajtodd reviewed (approved):
Fix and ship!
